etcdserver/etcdhttp/client_security.go

```diff
@@ -0,0 +1,399 @@
+	switch action {
+	case "grant":
+		out, err = sh.sec.UpdateRole(security.Role{Role: role, Grant: in.Grant})
+	case "revoke":
+		out, err = sh.sec.UpdateRole(security.Role{Role: role, Revoke: in.Revoke})
+	default:
+		return
+	}
+	if err != nil {
+		writeError(w, err)
+		return
+	}
+	err = json.NewEncoder(w).Encode(out)
+	if err != nil {
+		log.Println("etcdhttp: actionRole error encoding on", r.URL)
+	}
+}
+
+func (sh *securityHandler) baseUsers(w http.ResponseWriter, r *http.Request) {
+	if !allowMethod(w, r.Method, "GET") {
+		return
+	}
+	if !hasRootAccess(sh.sec, r) {
+		writeNoAuth(w)
+		return
+	}
+	w.Header().Set("X-Etcd-Cluster-ID", sh.clusterInfo.ID().String())
+	w.Header().Set("Content-Type", "application/json")
+	var usersCollections struct {
+		Users []string `json:"users"`
+	}
+	users, err := sh.sec.AllUsers()
+	if err != nil {
+		writeError(w, err)
+		return
+	}
+
+	usersCollections.Users = users
+	err = json.NewEncoder(w).Encode(usersCollections)
+	if err != nil {
+		log.Println("etcdhttp: baseUsers error encoding on", r.URL)
+	}
+}
+
+func (sh *securityHandler) handleUsers(w http.ResponseWriter, r *http.Request) {
+	subpath := path.Clean(r.URL.Path[len(securityPrefix):])
+	// Split "/users/username/command".
+	// First item is an empty string, second is "roles"
+	pieces := strings.Split(subpath, "/")[2:]
+	switch len(pieces) {
+	case 1:
+		sh.forUser(w, r, pieces[0])
+	case 2:
+		sh.actionUser(w, r, pieces[0], pieces[1])
+	}
+}
+
+func (sh *securityHandler) forUser(w http.ResponseWriter, r *http.Request, user string) {
+	if !allowMethod(w, r.Method, "GET", "PUT", "DELETE") {
+		return
+	}
+	if !hasRootAccess(sh.sec, r) {
+		writeNoAuth(w)
+		return
+	}
+	w.Header().Set("X-Etcd-Cluster-ID", sh.clusterInfo.ID().String())
+
+	switch r.Method {
+	case "GET":
+		w.Header().Set("Content-Type", "application/json")
+		u, err := sh.sec.GetUser(user)
+		if err != nil {
+			writeError(w, err)
+			return
+		}
+		u.Password = ""
+		err = json.NewEncoder(w).Encode(u)
+		if err != nil {
+			log.Println("etcdhttp: forUser error encoding on", r.URL)
+			return
+		}
+	case "PUT":
+		var u security.User
+		err := json.NewDecoder(r.Body).Decode(&u)
+		if err != nil {
+			writeError(w, httptypes.NewHTTPError(http.StatusBadRequest, err.Error()))
+			return
+		}
+		if u.User != user {
+			httptypes.NewHTTPError(400, "User JSON name does not match the name in the URL")
+		}
+		err = sh.sec.CreateUser(u)
+		if err != nil {
+			writeError(w, err)
+			return
+		}
+	case "DELETE":
+		err := sh.sec.DeleteUser(user)
+		if err != nil {
+			writeError(w, err)
+			return
+		}
+	}
+}
+
+func (sh *securityHandler) actionUser(w http.ResponseWriter, r *http.Request, user string, action string) {
+	if !allowMethod(w, r.Method, "PUT") {
+		return
+	}
+	if !hasRootAccess(sh.sec, r) {
+		writeNoAuth(w)
+		return
+	}
+	w.Header().Set("X-Etcd-Cluster-ID", sh.clusterInfo.ID().String())
+	w.Header().Set("Content-Type", "application/json")
+	var in security.User
```

Comment:
Shall we compare if in.User(if given) is equal to user? We actually duplicate the user information at two places (in path and json)

for example:

```
http://127.0.0.1:4001/v2/security/users/foo/password -XPUT -d '{"user":"bar", "password":"123"}'
```

should fail and give 

```
{"message": "Conflict user name in path and json"}
```